api/src/main/java/org/openmrs/module/fgh/mpi/processor/InitialLoadProcessor.java

```diff
@@ -0,0 +1,343 @@
+	
+	private static final int MPI_BATCH_SIZE = Integer.parseInt(MpiUtils.getGlobalPropertyValue(GP_INITIAL_BATCH_SIZE));
+	
+	private final ExecutorService executor = Executors.newFixedThreadPool(THREAD_COUNT);
+	
+	private final List<CompletableFuture<Map<String, Object>>> futures = synchronizedList(new ArrayList<>(THREAD_COUNT));
+	
+	private final AtomicInteger successCount = new AtomicInteger(0);
+	
+	private Long start;
+	
+	private List<Integer> lastSubmittedPatientIds = new ArrayList<>();
+	
+	private final PatientService patientService = Context.getPatientService();
+	
+	private final MpiHttpClient mpiHttpClient;
+	
+	private final ObjectMapper objectMapper = new ObjectMapper();
+	
+	private InitialLoadTaskStatus initialLoadTaskStatus;
+	
+	public InitialLoadProcessor(MpiHttpClient mpiHttpClient) {
+		super(true);
+		this.mpiHttpClient = mpiHttpClient;
+	}
+	
+	public void runInitialLoad() {
+		try {
+			initialLoadTaskStatus = MpiUtils.fetchInitialLoadTaskController();
+		}
+		catch (SQLException e) {
+			log.error("Error while fetching initial load task info", e);
+			throw new RuntimeException(e);
+		}
+		
+		if (initialLoadTaskStatus != null && initialLoadTaskStatus.isLocked()) {
+			log.info("Initial load task is already running");
+		} else {
+			
+			try {
+				MpiContext.initIfNecessary();
+			}
+			catch (Exception e) {
+				throw new APIException("Failed to initialize MpiContext", e);
+			}
+			
+			log.info("Starting Patient initial load process");
+			start = System.currentTimeMillis();
+			boolean continueProcessing = true;
+			
+			try {
+				while (continueProcessing) {
+					Integer lastId = this.initialLoadTaskStatus != null ? this.initialLoadTaskStatus.getPatientOffsetId()
+					        : 0;
+					Integer id = MpiUtils.getLastSubmittedPatientId() != null ? MpiUtils.getLastSubmittedPatientId()
+					        : lastId;
+					
+					List<Integer> patientsId = MpiUtils.executePatientQuery(
+					    MpiIntegrationProcessor.PATIENT_QUERY_BOUNDARIES.replace(ID_PLACEHOLDER, String.valueOf(id))
+					            .replace(MpiIntegrationProcessor.MAXIMUM_RESULT_PLACEHOLDER, String.valueOf(BATCH_SIZE)));
+					
+					log.info("Found {} Patients for initial load process", patientsId.size());
+					if (patientsId.isEmpty()
+					        || (patientsId.size() == 1 && patientsId.get(0).equals(MpiUtils.getLastSubmittedPatientId()))) {
+						if (this.initialLoadTaskStatus != null) {
+							this.initialLoadTaskStatus.setRunning(false);
+							this.initialLoadTaskStatus.setEndDate(new Date());
+							this.initialLoadTaskStatus.setActive(Boolean.FALSE);
+							this.initialLoadTaskStatus.setLocked(Boolean.FALSE);
+							MpiUtils.updateInitialLoadTaskController(this.initialLoadTaskStatus);
+						}
+						continueProcessing = false;
+						log.info("No patients found for initial load process");
+					} else {
+						
+						//Create a task controller
+						if (this.initialLoadTaskStatus == null) {
```

Comment:
You won't need to do this check if you always create "initialLoadTaskStatus" at beggining of the method

Reply:
Done